labgrid/resource/mqtt.py

```diff
@@ -0,0 +1,81 @@
+import queue
+import logging
+from time import monotonic
+
+import attr
+import paho.mqtt.client as mqtt
+import threading
+
+from .common import ManagedResource, ResourceManager, Resource
+from ..factory import target_factory
+from ..util import Timeout
+
+@attr.s(eq=False)
+class MQTTManager(ResourceManager):
+    _available = attr.ib(default=attr.Factory(list), validator=attr.validators.instance_of(list))
+    _avail_lock = attr.ib(default=threading.Lock())
+    _clients = attr.ib(default=attr.Factory(dict), validator=attr.validators.instance_of(dict))
+    _topics = attr.ib(default=attr.Factory(list), validator=attr.validators.instance_of(list))
+    _topic_lock = attr.ib(default=threading.Lock())
+    _last = attr.ib(default=0.0, validator=attr.validators.instance_of(float))
+
+    def __attrs_post_init__(self):
+        super().__attrs_post_init__()
+        self.log = logging.getLogger('MQTTManager')
+
+    def _on_connect(self, client, userdata, flags, rc):
+        client.subscribe("#")
+
+    def _create_mqtt_connection(self, host):
+        client = mqtt.Client()
+        client.connect(host)
+        client.on_connect = self._on_connect
+        client.on_message = self._on_message
+        client.loop_start()
+        return client
+
+    def on_resource_added(self, resource):
```

Comment:
Should this be private?

Reply:
This is called by the `ResourceManager` class, changing this will require changing all other classes using `ResourceManager` as well.

Reply:
Ah, yes.